### ISSUE TITLE
Cache bitfield to disk

### ISSUE THREAD

**Issue by mafintosh** [mafintosh/torrent-stream#27]
Currently every restart will result in all stored data being reverified.
We should cache the bitfield to disk once in while so this is not needed.

**Comment by xeoncore** [comment]
It's still needed to make sure the file hasn't been corrupted. I was thinking about this and doing "lazy hashing" only when the piece is requested by the file stream or a client, or when a piece is about to be downloaded.

**Comment by mafintosh** [comment]
Lazy verification probably won't work since you need to send the bitfield with your available pieces right away when you connect to another peer (which means you have to verify all pieces right away) :(

**Comment by xeoncore** [comment]
Not exactly. The way I've seen other implementations do it (which would work well for us considering TorrentStream's purpose is a streaming implementation rather than a "download" implementation) is they hash progressively as needed and send 'have' messages whenever they hash a new piece. Ideally what you'd do is this:

- Background verify the hashes sequentially, probably based on a queue of sorts.
- Like is currently implemented with the "critical" pieces in torrent-stream, prioritize certain pieces in the hashing pipeline when they are requested by the client (eg. from the filestorage stream). 

This means that in the overall scheme of things, the amount of time to fully hashcheck a torrent would be roughly the same as it is now but it would mean that critical pieces as requested by the storage stream would be hash checked first and could then be used. For example, a lot of video containers store crucial metadata at the end of and between streams meaning in order for you to stream those containers it would require the (currently sequential) hashchecker to fully hashcheck the entire file before it can fulfill the requests for the pieces near the end of the torrent.

With "lazy" hash checking, the storage stream could be available immediately, even before hash checking is complete, when a stream is requested it finds the piece at the offset requested and, if its been hash checked already either passes the data back or sends a request for the piece to the swarm.
